### PR TITLE
MTDSA-4319 added service for retrieving non-fhl property

### PR DESCRIPTION
**Pull request by david-walters-HMRC**

### CODE REVIEW

File: app/v1/services/SelfEmploymentPropertyBISSService.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import v1.models.errors._
+import v1.models.outcomes.ResponseWrapper
+import v1.models.requestData.RetrieveUKPropertyBISSRequest
+import v1.models.response.RetrieveUKPropertyBISSResponse
+import v1.support.DesResponseMappingSupport
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class SelfEmploymentPropertyBISSService @Inject()(connector: SelfEmploymentBISSPropertyConnector) extends DesResponseMappingSupport with Logging {
```
Comment (narendravijayarao) [inline app/v1/services/SelfEmploymentPropertyBISSService.scala]: Rename the file with UKPropertyBISSService

Comment (narendravijayarao) [reply app/v1/services/SelfEmploymentPropertyBISSService.scala]: Rename the connector as UKPropertyBISSConnector

File: test/v1/mocks/connectors/MockSelfEmploymentPropertyBISSConnector.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import org.scalamock.handlers.CallHandler
+import org.scalamock.scalatest.MockFactory
+import uk.gov.hmrc.http.HeaderCarrier
+import v1.connectors.{DesOutcome, SelfEmploymentBISSPropertyConnector}
+import v1.models.requestData.RetrieveUKPropertyBISSRequest
+import v1.models.response.RetrieveUKPropertyBISSResponse
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait MockSelfEmploymentPropertyBISSConnector extends MockFactory {
```
Comment (narendravijayarao) [inline test/v1/mocks/connectors/MockSelfEmploymentPropertyBISSConnector.scala]: Rename it as MockUKPropertyBISSConnector

File: test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    )
+
+  "retrieveBiss" should {
+    "return a valid response" when {
+      "a valid request is supplied" in new Test {
+        MockSelfEmploymentBISSPropertyConnector.retrieveBiss(requestData)
+          .returns(Future.successful(Right(ResponseWrapper(correlationId, response))))
+
+        await(service.retrieveBiss(requestData)) shouldBe Right(ResponseWrapper(correlationId, response))
+      }
```
Comment (narendravijayarao) [inline test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala]: Add another scenario with "only mandatory fields"

File: test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+          MockSelfEmploymentBISSPropertyConnector.retrieveBiss(requestData)
+            .returns(Future.successful(Left(ResponseWrapper(correlationId, DesErrors.single(DesErrorCode(desErrorCode))))))
+
+          await(service.retrieveBiss(requestData)) shouldBe Left(ErrorWrapper(Some(correlationId), error))
+        }
+
+      val input = Seq(
+
+        ("INVALID_IDVALUE", NinoFormatError),
+        ("INVALID_INCOMESOURCEID", DownstreamError),
```
Comment (narendravijayarao) [inline test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala]: Missing  "INVALID_TAXYEAR" -> TaxYearFormatError scenario... It's easy if we start with 400's, then 403 (if any), 404 and lastly 500's

File: test/v1/mocks/connectors/MockSelfEmploymentPropertyBISSConnector.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  val mockConnector: SelfEmploymentBISSPropertyConnector = mock[SelfEmploymentBISSPropertyConnector]
+
+  object MockSelfEmploymentBISSPropertyConnector {
```
Comment (Nboaram) [inline test/v1/mocks/connectors/MockSelfEmploymentPropertyBISSConnector.scala]: Rename Object to MockUKPropertyBISSPropertyConnector

File: test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+class SelfEmploymentPropertyBISSServiceSpec extends UnitSpec {
+
+  private val nino = "AA123456A"
+  private val taxYear = "2019"
+  private val selfEmploymentId = "123456789"
+  private val correlationId = "X-123"
+
+  private val requestData = RetrieveUKPropertyBISSRequest(Nino(nino), DesTaxYear(taxYear), IncomeSourceType.`uk-property`)
+
+  trait Test extends MockSelfEmploymentPropertyBISSConnector {
```
Comment (Nboaram) [inline test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala]: MockUKPropertyBISSConnector?

File: test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    implicit val hc: HeaderCarrier = HeaderCarrier()
+    implicit val logContext: EndpointLogContext = EndpointLogContext("controller", "retrieveSelfEmploymentPropertyBISS")
+
+    val service = new SelfEmploymentPropertyBISSService(mockConnector)
```
Comment (Nboaram) [inline test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala]: UKProperty?

File: test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        await(service.retrieveBiss(requestData)) shouldBe Right(ResponseWrapper(correlationId, response))
+      }
+    }
+
+    "return error response" when {
+
+      def serviceError(desErrorCode: String, error: MtdError): Unit =
+        s"a $desErrorCode error is returned from the service" in new Test {
+
+          MockSelfEmploymentBISSPropertyConnector.retrieveBiss(requestData)
```
Comment (Nboaram) [inline test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala]: UKProperty?

File: test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      Some(Loss(
+        net = Some(10.00),
+        taxable = Some(35.00)
+      ))
+    )
+
+  "retrieveBiss" should {
+    "return a valid response" when {
+      "a valid request is supplied" in new Test {
+        MockSelfEmploymentBISSPropertyConnector.retrieveBiss(requestData)
```
Comment (Nboaram) [inline test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala]: UKProperty?

**Pull request merged**